src/material-examples/tree-checklist/tree-checklist-example.ts

```diff
@@ -168,10 +168,21 @@ export class TreeChecklistExample {
     return flatNode;
   }
 
-  /** Whether all the descendants of the node are selected */
+  /**
+   * Whether all the descendants of the node are selected.
+   * And remove the node from this.checklistSelection if it isn't selected.
```

Comment:
I don't think that this function is the proper place to be doing the update. The function is used as a getter and it shouldn't be used to make changes to other properties. It might be more appropriate to do the change in the `todoItemSelectionToggle` method that is called when a checkbox is changed.

Reply:
I also thought about this; When toggling a child, it will need to find its parent and calculate its checked state. I'll refactor this after work today.

Reply:
Also when a child is toggled it calls the `this.checklistSelection.toggle(node)` directly, so I will have to make another function.

Reply:
There seems to not be a function to get the parent of a node. So there are two options: 

1. Create a getParents function in `FlatTreeControl`
2. Create a getParents function in `tree-checklist-example.ts`

For this PR I think changing only the `tree-checklist-example.ts` is the better way, then create a new Feature Request to implement getParents in `FlatTreeControl`

Reply:
I've been looking at this the whole day and thinking about a solution.

Firstly the problem lies in the fact that the isSelected state of parent nodes aren't the same value if you select it or select all the children. 

So my solution was to try and rectify this, and keep the parent nodes selected state consistent.

I agree with your statement above stating that `descendantsAllSelected` is a getter and should be a pure function and not mutate the state of the object, but because of the nature of nested (_even flattened_) structures it can potentially be a very expensive operation to update every parent up the tree structure (_recursive_), which what needs to happen if we are going to change the state of the parent nodes on the `(change)` function. 

So I propose my solution which tries to be a lightweight as possible (_no recursion_), but it does unfortunately mutate state in the getter. I have moved the mutating function to a new function.

Your thoughts would be greatly appreciated

Reply:
I don't think that recursing through the entire tree would be that expensive since it only happens once when the checkbox is toggled. Even if it's thousands of nodes, we're probably looking at something that's going to take a couple of milliseconds max. If we put aside the fact that the getter shouldn't be changing any state, there's still the issue that the getter is invoked for every change detection cycle, even if the user didn't interact with the tree.